File: Medium/079. WordSearch/079. WordSearch.java
class Solution {
    int ROW;
    int COL;

    int[] dir = {0, 1, 0, -1, 0};
    public boolean exist(char[][] board, String word) {
        ROW = board.length;
        COL = board[0].length;

        for(int r=0; r<ROW; r++){
            for(int c=0; c<COL; c++){
                if(dfs(board, word, r, c, 0)){
                    return true;
                }
            }
        }
        return false;
    /**
    check all cells
    if a cell matches firstletter try
        mark as visited
        check all directions for next letter
    
    */
}
    public boolean dfs(char[][] board, String word, int row, int col, int index){
        if(row >= ROW || row < 0 || col >= COL || col < 0 || word.charAt(index) != board[row][col]) return false;
        if(index == word.length()-1 && board[row][col] == word.charAt(index)) return true;
        
        board[row][col] = '!';
        for(int i=0; i<dir.length-1; i++){
            if(dfs(board, word, row + dir[i], col + dir[i+1], index + 1)) return true;
        }
        board[row][col] = word.charAt(index);
        return false;
    }
} 

File: Medium/079. WordSearch/079. WordSearch.py
from typing import List
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        #     right     left    up     down
        DIR = [(0,1), (0,-1), (1,0), (-1, 0)]
        r_max, c_max = len(board), len(board[0])
        def dfs(row, col, letter):

            if letter == len(word)-1 and board[row][col] == word[letter]: return True
            if board[row][col] != word[letter]: return False

            board[row][col] = '#' # -> mark cell as visited
            for dr, dc in DIR:
                r = row + dr
                c = col + dc

                if (r > -1 and r < r_max) and (c > -1 and c < c_max):
                    if dfs(r, c, letter + 1): return True
            
            board[row][col] = word[letter] # -> reset back to og so it can be revisited
            return False

        for r in range(r_max):
            for c in range(c_max):
                if dfs(r, c, 0):
                    return True
        return False

